hppc/src/main/templates/com/carrotsearch/hppc/KTypePgmIndex.java

```diff
@@ -0,0 +1,588 @@
+/*! #set($TemplateOptions.ignored = ($TemplateOptions.isKTypeAnyOf("GENERIC", "BYTE", "SHORT", "CHAR"))) !*/
+package com.carrotsearch.hppc;
+
+import com.carrotsearch.hppc.cursors.KTypeCursor;
+import com.carrotsearch.hppc.procedures.KTypeProcedure;
+
+import java.util.Arrays;
+import java.util.Iterator;
+
+/**
+ * Space-efficient index that enables fast rank/range search operations on a sorted sequence
+ * of <code>KType</code>.
+ * <p>Implementation of the PGM-Index described at
+ * <a href="https://pgm.di.unipi.it/">https://pgm.di.unipi.it/</a>, based on the paper
+ * <pre>
+ *   Paolo Ferragina and Giorgio Vinciguerra.
+ *   The PGM-index: a fully-dynamic compressed learned index with provable worst-case bounds.
+ *   PVLDB, 13(8): 1162-1175, 2020.
+ * </pre>
+ * It provides {@code rank} and {@code range} search operations.
+ * {@code indexOf()} is faster than B+Tree, and much more compact.
+ * {@code contains()} is between 4x to 7x slower than {@code IntHashSet#contains()}, but
+ * between 2.5x to 3x faster than {@link Arrays#binarySearch}.
+ * <p>
+ * Its compactness (40KB for 200MB of keys) makes it efficient for very large collections,
+ * the index fitting easily in the L2 cache. The {@code epsilon} parameter should be set
+ * according to the desired space-time trade-off. A smaller value makes the estimation more
+ * precise and the range smaller but at the cost of increased space usage. In practice,
+ * {@code epsilon} 32 or 64 is a good sweet spot.
+ * <p>
+ * Internally the index uses an optimal piecewise linear mapping from keys to their position
+ * in the sorted order. This mapping is represented as a sequence of linear models (segments)
+ * which are themselves recursively indexed by other piecewise linear mappings.
+ */
+/*! ${TemplateOptions.generatedAnnotation} !*/
+/*! #if ($templateonly) !*/ @SuppressWarnings({"rawtypes", "unchecked"}) /*! #end !*/
+public class KTypePgmIndex<KType> implements Accountable {
+
+  /** Empty immutable KTypePgmIndex. */
+  public static final KTypePgmIndex EMPTY = new KTypeEmptyPgmIndex();
+
+  /**
+   * Epsilon approximation range when searching the list of keys.
+   * Controls the size of the returned search range, strictly greater than 0.
+   * It should be set according to the desired space-time trade-off. A smaller value makes the
+   * estimation more precise and the range smaller but at the cost of increased space usage.
+   */
+  // With EPSILON=64: the benchmark with 200MB of keys shows that this PGM index requires
```

Comment:
shouldn't this be part of the javadoc? Seems like helpful information, even if it's a final setting.